zen/config.c

```diff
@@ -0,0 +1,99 @@
+#include "zen/config.h"
+
+#include <pwd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <toml.h>
+#include <unistd.h>
+
+#include "build-config.h"
+#include "zen-common/log.h"
+#include "zen-common/util.h"
+
+static struct toml_table_t *
+zn_config_get_toml_table(void)
+{
+  const char *config_dir_partial_path = "/.config";
+  const char *toml_file_partial_path = "/zen-desktop/config.toml";
+  char *config_home_env, *config_home;
+  config_home_env = getenv("XDG_CONFIG_HOME");
+  if (config_home_env == NULL) {
+    char *homedir_env = getenv("HOME");
+    if (homedir_env == NULL) homedir_env = getpwuid(getuid())->pw_dir;
+    if (homedir_env == NULL) {
+      zn_warn("Could not find the home directory");
+      return NULL;
+    }
+
+    config_home =
+        (char *)malloc(strlen(homedir_env) + strlen(config_dir_partial_path) +
+                       strlen(toml_file_partial_path) + 1);
+    sprintf(config_home, "%s%s%s", homedir_env, config_dir_partial_path,
+        toml_file_partial_path);
+  } else {
+    config_home = (char *)malloc(
+        strlen(config_home_env) + strlen(toml_file_partial_path) + 1);
+    sprintf(config_home, "%s%s", config_home_env, toml_file_partial_path);
+  }
+
+  FILE *fp = fopen(config_home, "r");
+  free(config_home);
+  if (fp == NULL) {
+    zn_warn("config.toml not found in %s\n", config_home);
```

Comment:
It could be for reasons other than the file not existing.

```suggestion
    zn_warn("Failed to open %s\n", config_home);
```